### PR TITLE
feat(cfw): add new EIP protection resource

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
support a new resource to protect one or more EIPs.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new EIP protection resource.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/cfw' TESTARGS='-run=TestAccEipProtection_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cfw -v -run=TestAccEipProtection_basic -timeout 360m -parallel 4
=== RUN   TestAccEipProtection_basic
=== PAUSE TestAccEipProtection_basic
=== CONT  TestAccEipProtection_basic
--- PASS: TestAccEipProtection_basic (100.22s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cfw       100.316s
```

### CODE REVIEW

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cfw' TESTARGS='-run=TestAccEipProtection_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cfw -v -run=TestAccEipProtection_basic -timeout 360m -parallel 4
=== RUN   TestAccEipProtection_basic
=== PAUSE TestAccEipProtection_basic
=== CONT  TestAccEipProtection_basic
--- PASS: TestAccEipProtection_basic (89.01s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cfw       89.097s
```

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cfw' TESTARGS='-run=TestAccEipProtection_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cfw -v -run=TestAccEipProtection_basic -timeout 360m -parallel 4
=== RUN   TestAccEipProtection_basic
=== PAUSE TestAccEipProtection_basic
=== CONT  TestAccEipProtection_basic
--- PASS: TestAccEipProtection_basic (105.00s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cfw       105.081s
```

File: docs/resources/cfw_eip_protection.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+---
+subcategory: "Cloud Firewall (CFW)"
+---
+
+# huaweicloud_cfw_eip_protection
+
+Manages a CFW protection rule resource within HuaweiCloud.
```
Comment (deer-hang) [inline docs/resources/cfw_eip_protection.md]:
```suggestion
Manages a CFW EIP protection resource within HuaweiCloud.
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+			"public_ip":   utils.ValueIngoreEmpty(raw["public_ipv4"]),
+			"public_ipv6": utils.ValueIngoreEmpty(raw["public_ipv6"]),
+		}
+		protectedEipIds[i] = raw["id"].(string)
+	}
+	return requestParams, protectedEipIds
+}
+
+// NoProtectedEips method check whether exist unprotected public IPs.
+func NoProtectedEips(eips []interface{}) bool {
```
Comment (deer-hang) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go]: How about rename this function name to ProtectedEipExist

Comment (Lance52259) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go:286]: NoProtectedEipsExist() maybe better.

Comment (Lance52259) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go]: seems be ok

Comment (Lance52259) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go]: done.

File: docs/resources/cfw_eip_protection.md (hunk 1 above)
Comment (Lance52259) [reply docs/resources/cfw_eip_protection.md]: done.

File: docs/resources/cfw_eip_protection.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource.
+  If omitted, the provider-level region will be used. Changing this parameter will create a new resource.
+
+* `object_id` - (Required, String, ForceNew) The protected object ID.
+  Changing this parameter will create a new resource.
+
+* `protected_eip` - (Required, List) The protected EIP configurations.
```
Comment (deer-hang) [inline docs/resources/cfw_eip_protection.md:47]: How about rename this field to `protected_eips`

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+				if eipId == utils.PathSearch("id", val, "").(string) &&
+					int(utils.PathSearch("status", val, "").(float64)) == targetStatus {
+					syncedEipCount++
+					break
+				}
+			}
+		}
+		if len(syncEips) > syncedEipCount {
+			return resp, "PENDING", nil
+		}
```
Comment (deer-hang) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go:176]: Comparing two arrays, it maybe better co compare array length first.

File: docs/resources/cfw_eip_protection.md (hunk 3 above)
Comment (Lance52259) [reply docs/resources/cfw_eip_protection.md:47]: `protected_eip` will be better

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+		rst = append(rst, map[string]interface{}{
+			"id":          utils.PathSearch("id", eip, nil),
+			"public_ipv4": utils.PathSearch("public_ip", eip, nil),
+			"public_ipv6": utils.PathSearch("public_ipv6", eip, nil),
+		})
+	}
+	return rst
+}
+
+func unprotectEips(ctx context.Context, client *golangsdk.ServiceClient, objectId string, unprotectedEipSet *schema.Set,
```
Comment (ShiChangkuo) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go]: it's better to re-arrange some functions, for example: put protectEips and unprotectEips together, put exportable functions together

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+		KeepResponseBody: true,
+		OkCodes:          []int{200},
+	}
+
+	path := client.Endpoint + modifyProtectHttpUrl
+	path = strings.ReplaceAll(path, "{project_id}", client.ProjectID)
+	opts.JSONBody = utils.RemoveNil(requestParams)
+	_, err := client.Request("POST", path, &opts)
+	if err != nil {
+		return fmt.Errorf("error closing EIP protection: %s", err)
```
Comment (ShiChangkuo) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go]: it's better to use enable/disable instead of open/close

Comment (Lance52259) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go]: enabling/disabling are already used

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_eip_protection.go]: done.

**Pull request merged**